keychain/interface_test.go

```diff
@@ -0,0 +1,255 @@
+					t.Fatalf("unable to derive key_index=%v "+
+						"for keyFam=%v: %v",
+						randKeyIndex, keyFam, err)
+				}
+			}
+		})
+		if !success {
+			break
+		}
+	}
+}
+
+// secretKeyRingConstructor is a function signature that's used as a generic
+// constructor for various implementations of the SecretKeyRing interface. A
+// string naming the returned interface, a function closure that cleans up any
+// resources, and the clean up interface itself are to be returned.
+type secretKeyRingConstructor func() (string, func(), SecretKeyRing, error)
+
+// TestSecretKeyRingDerivation tests that each known SecretKeyRing
+// implementation properly adheres to the expected behavior of the set of
+// interface.
+func TestSecretKeyRingDerivation(t *testing.T) {
+	t.Parallel()
+
+	secretKeyRingImplementations := []secretKeyRingConstructor{
+		func() (string, func(), SecretKeyRing, error) {
+			cleanUp, wallet, err := createTestBtcWallet()
+			if err != nil {
+				t.Fatalf("unable to create wallet: %v", err)
+			}
+
+			keyRing := NewBtcWalletKeyRing(wallet)
+
+			return "btcwallet", cleanUp, keyRing, nil
+		},
+	}
+
+	// For each implementation constructor registered above, we'll execute
+	// an identical set of tests in order to ensure that the interface
+	// adheres to our nominal specification.
+	for _, secretKeyRingConstructor := range secretKeyRingImplementations {
+		keyRingName, cleanUp, secretKeyRing, err := secretKeyRingConstructor()
+		if err != nil {
+			t.Fatalf("unable to create secret key ring %v: %v",
+				keyRingName, err)
+		}
+		defer cleanUp()
+
+		success := t.Run(fmt.Sprintf("%v", keyRingName), func(t *testing.T) {
+			// First, each key family, we'll ensure that we're able
+			// to obtain the private key of a randomly select child
+			// index within the key family.
+			for _, keyFam := range versionZeroKeyFamilies {
+				randKeyIndex := uint32(rand.Int31())
+				keyLoc := KeyLocator{
+					Family: keyFam,
+					Index:  randKeyIndex,
+				}
+
+				// First, we'll query for the public key for
+				// this target key locator.
+				pubKeyDesc, err := secretKeyRing.DeriveKey(keyLoc)
+				if err != nil {
+					t.Fatalf("unable to derive pubkey "+
+						"(fam=%v, index=%v): %v",
+						keyLoc.Family,
+						keyLoc.Index, err)
+				}
+
+				// With the public key derive, ensure that
+				// we're able to obtain the corresponding
+				// private key correctly.
+				privKey, err := secretKeyRing.DerivePrivKey(KeyDescriptor{
+					KeyLocator: keyLoc,
+				})
+				if err != nil {
+					t.Fatalf("unable to derive priv "+
+						"(fam=%v, index=%v): %v", keyLoc.Family,
+						keyLoc.Index, err)
+				}
+
+				// Finally, ensure that the keys match up
+				// properly.
+				if !pubKeyDesc.PubKey.IsEqual(privKey.PubKey()) {
+					t.Fatalf("pubkeys mismatched: expected %x, got %x",
+						pubKeyDesc.PubKey.SerializeCompressed(),
+						privKey.PubKey().SerializeCompressed())
+				}
+
+				// TODO(roasbeef): scalar mult once integrated
```

Comment:
Scalar mult impl is finished no? Should this TODO be addressed?

Reply:
Yes, but it isn't yet integrated into the rest of the code, as I started to, but then wanted to keep the set of changes in the PR relatively constrained.